Improve password hashing API

rushsteve1 opened the pull request:
I had been unhappy with the way that we interacted with the password hashing utilities for a while now so decided to finally get around to do something about it.

**Related Issues**
(I never made an issue for this, I should have)

**Describe the Changes**

This commit removes `gen_salt()` and merges it into `hash_password()` and introduces the `HashedPassword` struct which is used to represent a hashed password in a safer manner.

I also added a simple type def `UnsafeBinaryString` to make it more obvious that the output of a password hashing is not a valid string but we are treating it like one.

**Still To Do**
Ensure that the changes do not break any existing password hashes in the database. It should be fine since the actual password hashing and verifying code has not changed.

**Problems**
`HashedPassword` and `UnsafeBinaryString` were both introduced to make working with the API somewhat safer. But it still ends up using unsafe `String`s and uses unsafe code, which I would like to avoid.
It might be possible to have Diesel understand a `Vec<u8>` or similar and handle it in the database. But I do not know how.

Which is why I am making this as a PR instead of just pushing it like I usually do. Some feedback on the new API and any suggestions on tightening it up and possibly removing the unsafe code would be awesome.

So that said @BrianMejia would you be willing to review this for me?

rushsteve1 committed:
Improve password hashing API

I had been unhappy with the way that we interacted
with the password hashing utilities for a while now

This commit removes gen_salt() and merges it into hash_password()
and introduces the HashedPassword struct
which is used to represent a hashed password in a safer manner.
I also added a simple type def UnsafeBinaryString
to make it more obvious that the output of a password hashing
is not a valid string but we are treating it like one.

BrianMejia commented:
We've discussed a potential option using Diesel's serialization features. A good example comes from the [`ToSql`](https://docs.diesel.rs/diesel/serialize/trait.ToSql.html) trait, where we can use a `u8` implementation of this trait. We can then wrap the `UnsafeBinaryString` into a newtype struct, allowing the newtype to not be treated like a `String`.

A more drastic option is to just change the database backend (SQLite does not like to be complex), but of course that will take a lot of effort.

rushsteve1 committed:
Make UnsafeBinaryString a newtype struct

I remove the HashedPassword struct and replaced it with a more
robust UnsafeBinaryString struct that implements a derive
for all the Diesel types it needs to be compatible with the database.
With that all the calls to both() have been removed.

This should offer much more safety and generally be clearer.

rushsteve1 commented:
OK I have updated this to a new model that uses a custom type and makes Diesel understand it. I think this approach is much better, and generally provides the safety that I wanted out of this.

BrianMejia reviewed (approved):
After running locally, the hashing changes look like they don't affect current logins, and the newtype struct achieves a higher degree of safety. Looks great!
